File: src/onion/compiler/error/CompilationException.java
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005, Kota Mizushima, All rights reserved.       *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler.error;

import java.util.List;

/**
 * @author Kota Mizushima
 */
public class CompilationException extends RuntimeException {    
  private List problems;

  public CompilationException(List problems) {
    this.problems = problems;
  }
  
  public CompileError get(int index) {
    return (CompileError)problems.get(index);
  }
  
  public int size() {
    return problems.size();
  }
}

File: src/main/scala/onion/compiler/exceptions/CompilationException.scala
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2016-, Kota Mizushima, All rights reserved.  *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler.exceptions

import onion.compiler.CompileError

/**
 * @author Kota Mizushima
 * Exception thrown when compilation errors occurred.
 */
class CompilationException(val problems: Seq[CompileError])
    extends OnionException(CompilationException.formatMessage(problems))
    with Iterable[CompileError] {

  override def size: Int = problems.size

  override def iterator: Iterator[CompileError] = problems.iterator
}

object CompilationException {
  private def formatMessage(problems: Seq[CompileError]): String = {
    if (problems.isEmpty) "Compilation failed"
    else s"Compilation failed with ${problems.size} error(s)"
  }
}

